akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala

```diff
@@ -91,81 +125,173 @@ private[remote] final class TopHeavyHitters[T](val max: Int) {
   @tailrec private def replaceExistingHeavyHitter(foundHashIndex: Int, hashCode: HashCodeVal, item: T, count: Long): Boolean =
     if (foundHashIndex == -1) throw new NoSuchElementException(s"Item $item is not present in HeavyHitters, can not replace it!")
     else if (Objects.equals(items(foundHashIndex), item)) {
-      putCount(foundHashIndex, count) // we don't need to change `hashCode` or `item`, those remain the same
-      if (foundHashIndex == lowestHitterIdx) updateLowestHitterIdx() // need to update the lowestHitter since we just bumped its count 
-      false // not a "new" heavy hitter, since we only replaced it (so it was signaled as new once before) 
+      updateCount(foundHashIndex, count) // we don't need to change `hashCode` or `item`, those remain the same
+      fixHeap(heapIndex(foundHashIndex))
+      false // not a "new" heavy hitter, since we only replaced it (so it was signaled as new once before)
     } else replaceExistingHeavyHitter(findHashIdx(foundHashIndex + 1, hashCode), hashCode, item, count) // recurse
 
   private def findHashIdx(searchFromIndex: Int, hashCode: HashCodeVal): Int =
     findEqIndex(hashes, searchFromIndex, hashCode.get)
 
   /**
+   * Fix heap property on `heap` array
+   * @param index place to check and fix
+   */
+  @tailrec
+  private def fixHeap(index: Int): Unit = {
```

Comment:
I haven't got any free time until weekend. I will try to improve this on Saturday. 

@ktoso 
What do you think for pair-programming session on weekend?